Jetpack Boost: Image guide improvements

pyronaur opened the pull request:
#### Changes proposed in this Pull Request:
 * Image guide now takes pixel density into account
 * Adjusted image size severity
 * UI Improvements
 * Image Guide now responds to changing `currentSrc` property after the window is resized
 * Reduced the amount of window resize listeners
 * Moved image dimension data to Svelte stores
 * Refactored image size gathering
 * Added `jb-debug-ig` GET param to allow testing image guide without logging in
 * Handle situations when the image is smaller than expected
 * Added loading indicators while fetching image size information
 
#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
pc9hqz-1pa-p2

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
Apply patch, install packages, compile and play around with the image guide.

pyronaur committed:
[not verified] Take pixel density into account

pyronaur committed:
[not verified] Adjust severity

pyronaur committed:
[not verified] Tweak the smaller bubble font size

pyronaur committed:
[not verified] Improve the IG messaging around image sizes

pyronaur committed:
[not verified] Make it easier to debug without logging in while in feature flag mode

pyronaur committed:
[not verified] phpcs doesn't like whitespace

pyronaur committed:
[not verified] Fix types

pyronaur committed:
[not verified] Make it easier to debug

Temporarily disable production builds for Image Guide

pyronaur committed:
[not verified] Mark Observer as unused for now

pyronaur committed:
[not verified] Add a loading spinner

pyronaur committed:
[not verified] A little refactor

Everything in the UI is now powered by stores so that UI components always have access to the latest information.

This was needed to account for the asynchronous nature of fetching image weights and resizing screens. For example, the image URL can change after resizing the screen (because the currentSrc property can be dynamic) and so all the file weight data needs to be fetched again.

This also adds a lot of improvements in the amount event listeners are attached and the UI deals with asynchronous data flow much better.

pyronaur committed:
[not verified] Reformat files

pyronaur committed:
[not verified] ImageGuide -> Popup

pyronaur committed:
[not verified] Clean up Main.svelte

pyronaur committed:
[not verified] Improve how the loading state is handled in the pop-up

pyronaur committed:
[not verified] Update image dimensions asap

pyronaur committed:
[not verified] Prevent the Popup component from jumping around

pyronaur committed:
[not verified] Clarify image file dimensions label

pyronaur committed:
[not verified] Consistent class names

pyronaur committed:
[not verified] Group dynamic values together

pyronaur committed:
[not verified] Remove unused shadow

pyronaur committed:
[not verified] Make store files easier to find,move to stores dir

pyronaur committed:
[not verified] Clean up image source gathering

pyronaur committed:
[not verified] Simplify storing dimensions using derived stores

pyronaur committed:
[not verified] Prefix unused file with _unused

pyronaur committed:
[not verified] Add comments & clean up

pyronaur committed:
[not verified] Changelog entry

pyronaur committed:
[not verified] File Size information is 0 when it's unavailable

pyronaur committed:
[not verified] ES Lint Fixes

haqadn commented on projects/plugins/boost/app/features/image-guide/src/find-image-elements.ts:
```diff
@@ -0,0 +1,109 @@
+	// Intentionally not using an allow-list because images may
+	// be served from weird URLs like /images/1234?size=large
+	if ( value.startsWith( '/' ) ) {
+		return value.endsWith( '.svg' );
```
This may cause issues if the URL has a query param, e.g., `sample.svg?foo=bar`. Moreover, relative URLs without an initial slash are also possible, which will be missed here. 

But, I guess both cases are unlikely to appear in practice. So, I think this is fine.

pyronaur replied:
Oh, this might be an issue actually. I'll update with a check for query strings.

pyronaur committed:
[not verified] Run eslint fix

pyronaur committed:
[not verified] Update Svelte

pyronaur committed:
[not verified] Fix svelte warning

pyronaur committed:
[not verified] Fix import order warnings

pyronaur committed:
[not verified] Fix ESLint in Svelte

Turns out I needed to set my workspace settings to run ESLint in svelte with

```
		"eslint.validate": [
			"javascript",
			"typescript",
			"svelte"
		]
```

haqadn commented on projects/plugins/boost/app/features/image-guide/src/index.ts:
```diff
@@ -1,11 +1,32 @@
 
 /**
- * Initialize the admin bar toggle.
+ * A helper function to filter out images
+ * that are too small, for example
+ * avatars, icons, etc.
+ *
+ * @param  images An array of images to filter
+ */
+function discardSmallImages( images: MeasurableImage[] ) {
+	const minSize = 65;
+	const elements = images.filter( image => {
+		const { width, height } = image.getSizeOnPage();
+		if ( ! width || ! height ) {
+			return false;
+		}
+		return width >= minSize && height >= minSize;
+	} );
+
+	return elements;
+}
```
I think the dimensions of the actual file should be considered instead of using `getSizeOnPage()` to determine if an image should be skipped. What if a 50x50 avatar is loading a 1000x1000 image?

pyronaur replied:
> What if a 50x50 avatar is loading a 1000x1000 image?

That is a fair point, but I don't think it's something that it's something to address in this PR. We can cover edge cases like this in the cloud instead. For viewing reports on the front-end, I'd say 100x100 is actually the bare minimum to make it a comfortable size.

haqadn replied:
Sure let's do it in a separate PR. But, I think this is very important and probably the whole purpose of the image guide IMHO. I have seen the example case so many times in real world.

haqadn commented on projects/plugins/boost/app/features/image-guide/src/stores/MeasurableImageStore.ts:
```diff
@@ -0,0 +1,91 @@
+		} );
+	}
+
+	private deriveExpectedSize() {
+		return derived( this.sizeOnPage, sizeOnPage => {
+			return this.image.getExpectedSize( sizeOnPage );
+		} );
+	}
```
It wasn't clear immediately that this method provides the expected dimensions considering the DPR. Please add a docblock.

haqadn commented on projects/plugins/boost/app/features/image-guide/src/ui/Popup.svelte:
```diff
@@ -0,0 +1,231 @@
+<script lang="ts">
```
Not a strong opinion. Would it benefit from a different name as Popup has a different meaning? Maybe something like `GuideCard`.

pyronaur replied:
I was calling this `ImageGuide` previously and changed it to `Popup` because it's when jumping files int he image-guide project it's much easier to remember "I'm looking for a popup" rather than "I want to edit the pop-up called ImageGuide now". Naming things is hard, I think `GuideCard` is also not as useful. I could settle for `GuidePopover`, but, personally, I think `Popup` and `Bubble` components standing next to each other explain themselves pretty well.

pyronaur committed:
[not verified] Added a comment explaining expected size

pyronaur committed:
[not verified] Remove query string params before checking for svg

haqadn reviewed (dismissed):
👍🏼 🚢
